File: workspaces/aware_network/libs/comms/python/aware_comms/duplex/ipc/codec.py
"""Line-oriented frame codec for local IPC transports."""

from __future__ import annotations

import json
from typing import cast

from aware_comms.duplex.protocol import DuplexMessageFrame


class DuplexIpcFrameCodec:
    """Encode and decode newline-delimited duplex frames."""

    @staticmethod
    def encode_frame(frame: DuplexMessageFrame) -> bytes:
        return (frame.model_dump_json() + "\n").encode("utf-8")

    @staticmethod
    def decode_frame(payload: bytes | str) -> DuplexMessageFrame:
        raw = payload.decode("utf-8") if isinstance(payload, bytes) else payload
        line = raw.strip()
        if not line:
            raise ValueError("IPC frame payload is empty")
        data = cast(object, json.loads(line))
        if not isinstance(data, dict):
            raise ValueError("IPC frame payload must decode to an object")
        return DuplexMessageFrame.model_validate(data)


__all__ = ["DuplexIpcFrameCodec"]

File: workspaces/aware_network/libs/comms/dart/lib/service/duplex/ipc/codec.dart
import 'dart:convert';

import 'package:aware_comms/service/duplex/protocol/models.dart';

class DuplexIpcFrameCodec {
  const DuplexIpcFrameCodec();

  String encodeFrame(DuplexMessageFrame frame) {
    return '${jsonEncode(frame.toJson())}\n';
  }

  DuplexMessageFrame decodeFrame(String payload) {
    final line = payload.trim();
    if (line.isEmpty) {
      throw ArgumentError.value(
          payload, 'payload', 'IPC frame payload is empty');
    }
    final decoded = jsonDecode(line);
    if (decoded is! Map<String, dynamic>) {
      throw ArgumentError.value(
        payload,
        'payload',
        'IPC frame payload must decode to an object',
      );
    }
    return DuplexMessageFrame.fromJson(decoded);
  }
}
